pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1140,11 +1140,46 @@ type WorkflowStep struct {
 	// OnExit is a template reference which is invoked at the end of the
 	// template, irrespective of the success, failure, or error of the
 	// primary template.
+	// DEPRECATED: Use LifecycleHook.template instead.
 	OnExit string `json:"onExit,omitempty" protobuf:"bytes,11,opt,name=onExit"`
+
+	// Hooks holds exit hook which is invoked at the end of the
+	// template, irrespective of the success, failure, or error status of the primary template
+	Hooks *LifecycleHooks `json:"hooks,omitempty" protobuf:"bytes,12,opt,name=hooks"`
+}
+
+type LifecycleHooks struct {
+	// Exit hold the template and arguments which is invoked at the end of the
+	// workflow, irrespective of the success, failure, or error status of the primary template
+	Exit *LifecycleHook `json:"exit," protobuf:"bytes,1,opt,name=exit"`
```

Comment:
I think you can further simplify, as mandatory structs do not (should not) be pointer type:

```
	Exit LifecycleHook `json:"exit" protobuf:"bytes,1,opt,name=exit"`
```

Great comments on this.